frontend/src/components/Util/TaskEditors/TaskEditor/task-diff-reducer.ts

```diff
@@ -74,100 +54,45 @@ function initializer([mainTask, subTasks]: [MainTask, readonly SubTask[]]): Stat
 function reducer(state: State, action: Action): State {
   switch (action.type) {
     case 'EDIT_MAIN_TASK': {
-      const { mainTask, diff, ...restState } = state;
+      const { taskData, diff, ...restState } = state;
       const { change } = action;
       const newDiff = { ...diff, mainTaskEdits: { ...diff.mainTaskEdits, ...change } };
-      return { ...restState, mainTask: { ...mainTask, ...change }, diff: newDiff };
-    }
-    case 'ADD_SUBTASK': {
-      const { subTasks, diff, ...restState } = state;
-      const { newSubTask } = action;
-      const id = getNewSubTaskId();
-      const newSubTasks = [...subTasks, { id, ...newSubTask }];
-      const newDiff = {
-        ...diff,
-        subTaskCreations: diff.subTaskCreations.set(id, newSubTask),
-      };
-      return { ...restState, subTasks: newSubTasks, diff: newDiff };
-    }
```

Comment:
here is my hypothesis on what might go wrong.
you can see in the reducer that we create a new state based on the previous state, however, in your code, you update the state NOT based on the old state, but based on something in the prop.
This is potentially very problematic, because setState is async, and react may choose to batch update. When react batch updates, then any code that doesn't update based on previous state (aka. not using update function but directly give a new value) can mess up.

to solve the problem, you think you need to give up the approach that deep compares subtask content. Instead, you should compare using order, since we know for sure order is unique and it's immutable!